include/arc_utilities/thin_plate_spline.hpp

```diff
@@ -0,0 +1,167 @@
+        }
+
+        /**
+         * @note Checking for the vailidty of the input data is put off until
+         *  the warping function is solved
+         */
+        template <typename Derived>
+        void setTemplatePoints(
+                const Eigen::MatrixBase<Derived>& template_points)
+        {
+            using namespace Eigen;
+            static_assert(MatrixBase<Derived>::RowsAtCompileTime == DIMENSIONS ||
+                          MatrixBase<Derived>::RowsAtCompileTime == Dynamic,
+                          "INVALID DATA DIMENSIONS FOR template_points");
+
+            homogeneous_template_points_.resize(DIMENSIONS + 1, template_points.cols());
+            homogeneous_template_points_ << template_points, RowVectorXd::Ones(template_points.cols());
+            solved_ = false;
+        }
+
+        /**
+         * @note Checking for the vailidty of the input data is put off until
+         *  the warping function is solved
+         */
+        template <typename Derived>
+        void setTargetPoints(
+                const Eigen::MatrixBase<Derived>& target_points)
+        {
+            using namespace Eigen;
+            static_assert(MatrixBase<Derived>::RowsAtCompileTime == DIMENSIONS ||
+                          MatrixBase<Derived>::RowsAtCompileTime == Dynamic,
+                          "INVALID DATA DIMENSIONS FOR target_points");
+
+            target_points_ = target_points;
+            solved_ = false;
+        }
+
+        void solveForCoeffs()
+        {
+            using namespace Eigen;
+            if (homogeneous_template_points_.cols() != target_points_.cols())
+            {
+                throw_arc_exception(
+                            std::invalid_argument,
+                            "template and target points must have the same number of points");
+            }
+
+            const auto NUMPOINTS = homogeneous_template_points_.cols();
+
+            // Calculate the kernel matrix for the template (control) points
+            const MatrixXd Rsquared =
+                    EigenHelpers::CalculateSquaredDistanceMatrix(homogeneous_template_points_);
+            const MatrixXd kernel = Rsquared.unaryExpr(&ThinPlateSpline::KernelFunction);
+
+            // Form the system:   A * params = B
+            //       A = [K,       template'
+            //            template,    0]
+            //       B = [target'
+            //              0    ]
+            MatrixXd A(NUMPOINTS + DIMENSIONS + 1, NUMPOINTS + DIMENSIONS + 1);
+            A << kernel,                       homogeneous_template_points_.transpose(),
+                 homogeneous_template_points_, MatrixXd::Zero(DIMENSIONS + 1, DIMENSIONS + 1);
+            MatrixXd B(NUMPOINTS + DIMENSIONS + 1, DIMENSIONS);
+            B << target_points_.transpose(),
+                 MatrixXd::Zero(DIMENSIONS + 1, DIMENSIONS);
+
+            // params = A \ B
+            const Matrix<double, Dynamic, DIMENSIONS> params =
```

Comment:
I'm assuming this math is right, since I don't know the ThisPlateSpline algorithm